gapis/api/templates/vulkan_gfx_api_extras.tmpl

```diff
@@ -76,66 +77,144 @@
+    // Try to enable the validation layer first, see if we can create an
+    // VkInstance successfully.
+    if (!has_validation && addValidationLayer) {
+        layers[layer_count++] = validationLayerName;
+    }
+    pInfo->enabledLayerCount = layer_count;
+
+    if (!has_debug_report && addValidationLayer) {
+      exts[ext_count++] = debugReportName;
+    }
+    pInfo->enabledExtensionCount = ext_count;
+    
+    // Push to the stack, so the acutal function for calling
     // vkCreateInstance can use it.
-    stack->push(pCreateInfo);
+    stack->push(pInfo);
     stack->push(pAllocator);
     stack->push(pInstance);
 
-    return callVkCreateInstance(~0, stack, pushReturn);
+    if (callVkCreateInstance(~0, stack, true)) {
+      *result = stack->pop<uint32_t>();
+      return true;
+    }
+    *result = VkResult::VK_ERROR_INITIALIZATION_FAILED;
+    return false;
   }
 ¶
-  bool Vulkan::replayCreateVkDevice(Stack* stack, bool pushReturn) {
-    auto pDevice = stack->pop<VkDevice*>();
-    auto pAllocator = stack->pop<VkAllocationCallbacks*>();
-    auto pCreateInfo = stack->pop<VkDeviceCreateInfo*>();
+  bool Vulkan::replayCreateVkDevice(
+    Stack* stack, size_val physicalDevice,
+    const VkDeviceCreateInfo* pCreateInfo,
+    VkAllocationCallbacks* pAllocator, VkDevice* pDevice,
+    bool addValidationLayer, uint32_t* result) {
+
+    auto create_info = *pCreateInfo;
+    VkDeviceCreateInfo* pInfo = &create_info;
+
+    char validationLayerName[] = "VK_LAYER_LUNARG_standard_validation";
+    char debugReportName[] = "VK_EXT_debug_report";
 
+    auto layer_count = pInfo->enabledLayerCount;
+    auto ext_count = pInfo->enabledExtensionCount;
+    const char** layers = static_cast<const char**>(alloca((layer_count + 1) * sizeof(char*)));
+    const char** exts = static_cast<const char**>(alloca((ext_count + 1) * sizeof(char*)));
+
+    bool has_validation = false;
+    for (size_t i = 0; i < layer_count; i++) {
+      layers[i] = pInfo->ppEnabledLayerNames[i];
+      if (strcmp(validationLayerName, pInfo->ppEnabledLayerNames[i]) == 0) {
+        has_validation = true;
+      }
+    }
+    pInfo->ppEnabledLayerNames = layers;
+
+    bool has_debug_report = false;
+    for (size_t i = 0; i < ext_count; i++) {
+      exts[i] = pInfo->ppEnabledExtensionNames[i];
+      if (strcmp(debugReportName, pInfo->ppEnabledExtensionNames[i]) == 0) {
+        has_debug_report = true;
+      }
+    }
+    pInfo->ppEnabledExtensionNames = exts;
     // We recorded pNext during capturing since we turned on VkGraphicsSpy layer.
     // On the replay side, since there are no extensions that can use pNext yet,
     // null it out.
-    pCreateInfo->pNext = nullptr;
+    pInfo->pNext = nullptr;
 
+    if (!has_validation && addValidationLayer) {
+      layers[layer_count++] = validationLayerName;
+    }
+    pInfo->enabledLayerCount = layer_count;
+    if (!has_debug_report && addValidationLayer) {
+      exts[ext_count++] = debugReportName;
```

Comment:
VK_EXT_debug_report is an instance extension so we should not enable it for device. This code is just added here to test the fallback route. Will be removed later in this PR.